records/2023-01-24.malicious-content-scanning.md

```diff
@@ -0,0 +1,104 @@
+
+We are now calling the scan-files api directly from the Document Download API ([Draft PR](https://github.com/cds-snc/notification-document-download-api/pull/97)). This API can take a few seconds to return a response, and this time increases as file size increases. So we need to stop this from causing delaying the response when a user makes a requests to the Notify API.  
+
+We have explored using the `asyncio` library in python to create a new process to call the scan-files api [here](https://github.com/cds-snc/notification-document-download-api/pull/97/files#diff-a842b15bd72bea41437e64bda92d12262c8eca35af7e8adf111bd520fe7c3010R1). Then once the result is returned, we could write that result to the S3 object as a tag which can be read later. However, this approach did not work to reduce the response time for the Document Download API.
+
+It seems that it is [not possible to create background processes inside flask](https://flask.palletsprojects.com/en/2.2.x/async-await/#background-tasks).
+
+## Options
+
+Here are some of the options for scanning email attachments with the scan-files project.
+
+### Duplicate S3 bucket
+
+The scan-files project comes with a terraform module that triggers a scan of newly created S3 objects in a bucket it is told to watch. The S3 bucket we are currently using to store file attachments won't work, since these file are each encrypted with a unique key and the scan-files tf module does not support this. So, we could create a 2nd S3 bucket with restricted access containing duplicates of these files, all encrypted with the same key.
+
+Mapping files from one S3 bucket to the other is simple - we can just use the same file name in both cases. We would need to enforce retention/deletion policies in the duplicate bucket that are at least as strict. And the duplicate S3 bucket can have an access policy that restricts access to the document-download API and scan-files.
+
+We only need to store the duplicate files until the malware scan is complete and the scan verdict is read by notification-api prior to sending the email and file attachment (either in linked or attached directly).
+
+Here is what happens when an API user makes a request to send an email with an attachment:
+
+```mermaid
+flowchart TD
+    A[user makes request to /email with file attachment]
+    B[notification-api uploads file to dd-api]
+    C[dd-api uploads file to:\n document-download S3 bucket \n scan-files S3 bucket]
+    D[dd-api returns a 201 to notification-api]
+    E[notification-api returns a 201 to user, sets notification status to 'Sending']
+    A-->B
+    B-->C
+    C-->D
+    D-->E
+```
+Eventually the email sending job is processed by Celery, and the following occurs: 
+```mermaid
+flowchart TD
+    A[Celery calls `send_email_to_provider` fn]
+    B{Does email contain a file attachment?}
+    C1{request to dd-api to \n/check-scan-verdict endpoint}
+    C2[Send email as usual]
+    D[Wait for 5 minutes and try again]
```

Comment:
I realize that I'm the one that steered you down the path towards async scanning, but that may have been a mistake if you're going to delay the email send until a verdict is returned.  Apologies for not realizing this sooner, but if you used a synchronous invoke of the Scan Files API from the Notify code, it would accomplish the same thing.

That being said, the risk would be if there's a problem with the Scan Files API that prevents it from scanning, so some timeout logic would be needed on the API call.

Reply:
That's okay! I think the S3 bucket solution is still better for us. If we called the scan-files api directly and waited for a response, that wait would have to happen in Celery, so that we don't take minutes to return an API response to the user. And then in Celery, the waiting for a response could have worse performance that making a request to "/check-scan-verdict" on the dd-api, quickly getting the response, and then trying again later if the scan is still in progress.

Though I am now thinking that retrying after 5 minutes may be too long. If the scan is often less than 1 minute, we could try again at 1 minute, 2 minutes, etc

Reply:
Gotcha and yeah, most responses are returned within a second or two.  It's only in the case of a cold-start on our scanning API that we see the response take ~30 seconds.

Reply:
@smcmurtry 💯 We like the async nature of it so we don't block on process or pollute it.